runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,341 @@
+
+local document_color_ns = api.nvim_create_namespace('nvim.lsp.document_color')
+local document_color_augroup = api.nvim_create_augroup('nvim.lsp.document_color', {})
+
+local M = {}
+
+--- @alias vim.lsp.document_color.HighlightInfo { hex_code: string, hl_group?: string, range: Range4 }
+
+--- @class (private) vim.lsp.document_color.GlobalState
+--- @field enabled boolean
+local global_state = { enabled = false }
+
+--- @class (private) vim.lsp.document_color.BufState : vim.lsp.document_color.GlobalState
+--- @field buf_version? integer Buffer version for which the color ranges correspond to.
+--- @field applied_version? integer Last buffer version for which we applied color ranges.
+--- @field hl_infos? table<integer, vim.lsp.document_color.HighlightInfo[]> client_id -> processed color highlights
+
+--- @type table<integer, vim.lsp.document_color.BufState>
+local bufstates = vim.defaulttable(function(_)
+  return setmetatable({}, {
+    __index = global_state,
+    __newindex = function(state, key, value)
+      if global_state[key] == value then
+        rawset(state, key, nil)
+      else
+        rawset(state, key, value)
+      end
+    end,
+  })
+end)
+
+--- @inlinedoc
+--- @class vim.lsp.document_color.enable.Opts
+---
+--- Highlight style. It can be one of the pre-defined styles or a function that receives the buffer handle,
+--- the range (start line, start col, end line, end col) and the resolved hex color.
+--- Defaults to 'background'.
+--- @field style? 'foreground'|'background'|'virtual'|fun(bufnr: integer, range: Range4, hex_code: string)
+
+-- Default options.
+--- @type vim.lsp.document_color.enable.Opts
+M._opts = { style = 'background' }
+
+--- @param color string
+local function get_contrast_color(color)
+  local r_s, g_s, b_s = color:match('^#(%x%x)(%x%x)(%x%x)$')
+  local r, g, b = tonumber(r_s, 16), tonumber(g_s, 16), tonumber(b_s, 16)
+
+  local luminance = 0.298912 * r + 0.586611 * g + 0.114478 * b
+  local is_bright_color = luminance > 127
+  return is_bright_color and '#000000' or '#ffffff'
+end
+
+--- Returns the hex string representing the given LSP color.
+--- @param color lsp.Color
+--- @return string
+local function get_hex_code(color)
+  -- The RGB values in lsp.Color are in the [0-1] range, but we want them to be in the [0-255] range instead.
+  --- @param n number
+  color = vim.tbl_map(function(n)
+    return math.floor((n * 255) + 0.5)
+  end, color)
+
+  return ('#%02x%02x%02x'):format(color.red, color.green, color.blue):lower()
+end
+
+--- Cache of the highlight groups that we've already created.
+--- @type table<string, true>
+local color_cache = {}
+
+--- Gets or creates the highlight group for the given LSP color information.
+---
+--- @param hex_code string
+--- @param style 'foreground'|'background'|'virtual'
+--- @return string
+local function get_hl_group(hex_code, style)
+  local hl_name = ('LspDocumentColor_%s_%s'):format(hex_code:sub(2), style)
+
+  if not color_cache[hl_name] then
```

Comment:
Highlight groups can be cleared after they are created. The common use case is usually `:colorscheme` command, but can happen after a manual `:hi clear` and intentional `vim.api.nvim_set_hl(0, 'LspDocumentColor_000000_background', {})` (which is not reasonable to support, but still can happen).

There are several ways around it:
- `vim.fn.hlexists(hl_name) == 0`. Very fast, but might have issues with fast events (needs checking).
- `next(vim.api.nvim_get_hl(0, { name = hl_name })) == nil`. Equally fast, not sure about side effects. Kind of cryptic.
- Clearing `color_cache` in autocommand for `ColorScheme` event. Not 100% solution, but covers most use cases.

I'd say `nvim_get_hl()` is the best overall approach here (if works).

Reply:
Hmm I think I'll go with just reacting to the `ColorScheme` event ([like `mini.hipatterns` does](https://github.com/echasnovski/mini.hipatterns/blob/e5083df391171dc9d8172645606f8496d9443374/lua/mini/hipatterns.lua#L707) 😉)

Reply:
Yeah, this works, but mostly because it is also a convenient way to reload buffers. Although, after writing the original comment, I did start to think that using `vim.api.nvim_get_hl()` in 'mini.hipatterns' instead of custom cache might be worth it.

Reply:
Okay, I'll switch it over.

Reply:
> Okay, I'll switch it over.

Sorry for not being clear. That comment was mostly "thinking out loud". Using `ColorScheme` event definitely works and should be enough here (the only good reason to call `:hi clear` is inside a color scheme file).